### PR TITLE
Support reduceMean for IE/DML

### PR DESCRIPTION
**Pull request by mingmingtasd**
PTAL, thanks! @fujunwei

### CODE REVIEW

File: src/webnn_native/openvino/ienn/src/ie_model.cpp [hunk 1]
```diff
@@ -438,6 +438,21 @@ ie_operand_t* Model::AddRelu(ie_operand_t* input) {
...
 
+ie_operand_t* Model::AddReduceMean(ie_operand_t* input,
+                                   ie_reduce_mean_options_t* options) {
+  auto input_node = name_node_map_[input->name];
+  SizeVector axes = ToVector(options->axes, options->axesCount);
+  auto axes_node = op::Constant::create(element::i64, Shape{axes.size()}, axes);
+  // Issue: When input rank == 3 and keeDimensions == false, reduceMean produces
+  // a result with wrong shape. This will be fixed by
+  // https://github.com/openvinotoolkit/openvino/pull/5986 in OpenVINO 2021/4.
+  auto reduce_mean_node = std::make_shared<op::v1::ReduceMean>(
```
Comment (fujunwei) [inline src/webnn_native/openvino/ienn/src/ie_model.cpp:449]: https://docs.openvinotoolkit.org/latest/openvino_docs_ops_reduction_ReduceMean_1.html
> Scalar or 1D tensor of type T_IND with axis indices for the 1st input along which reduction is performed. Accepted range is [-r, r-1] where where r is the rank of input tensor.

-1 is not the last dimension?

> all values must be unique, repeats are not allowed. Required.

Maybe we need to verify it, and double check it for DML.

File: src/webnn_native/dml/GraphDML.cpp [hunk 2]
```diff
@@ -874,6 +874,43 @@ namespace webnn_native { namespace dml {
...
+        DAWN_ASSERT(mExpression.find(inputOperand) != mExpression.end());
+        ::dml::Expression input = mExpression.at(inputOperand);
+        const ReduceMeanOptions* options = reduceMean->GetOptions();
+        const int32_t* axesData = reinterpret_cast<const int32_t*>(options->axes);
+        std::vector<std::uint32_t> axesVector;
+        size_t inputRank = input.GetOutputDesc().sizes.size();
+        for (size_t i = 0; i < options->axesCount; ++i) {
+            // Axes values must be in the range [0, InputTensor.DimensionCount - 1].
+            // The dimensions to reduce where -1 means the last dimension.
+            uint32_t axis = axesData[i] == -1 ? inputRank - 1 : axesData[i];
```
Comment (fujunwei) [inline src/webnn_native/dml/GraphDML.cpp:889]: Do we support multiply -1 in the axes?

File: src/webnn_native/dml/GraphDML.cpp [hunk 3]
```diff
@@ -874,6 +874,43 @@ namespace webnn_native { namespace dml {
...
+            axesVector.push_back(axis);
+        }
+        ::dml::Span<const uint32_t> axes(axesVector);
+        ::dml::Expression output = ::dml::Reduce(input, DML_REDUCE_FUNCTION_AVERAGE, axes);
+        ::dml::TensorDimensions outputDims = output.GetOutputDesc().sizes;
+        if (!options->keepDimensions) {
+            ::dml::TensorDimensions newDims;
+            for (size_t i = 0; i < outputDims.size(); ++i) {
+                if (!(outputDims[i] == 1 && std::find(axes.begin(), axes.end(), i) != axes.end())) {
+                    newDims.push_back(outputDims[i]);
```
Comment (fujunwei) [inline src/webnn_native/dml/GraphDML.cpp:899]: How about removing the dimension in the `outputDims `with the axes?

**Review by fujunwei — COMMENTED** [review]
I have some comments, please take a look, thanks.

File: src/webnn_native/openvino/ienn/src/ie_model.cpp (hunk 1 above)
Comment (mingmingtasd) [reply src/webnn_native/openvino/ienn/src/ie_model.cpp:449]: -1 is the last dimension in IE. And DML dosen't support -1 as last dimension, I have converted it when it's -1 at this [line](https://github.com/webmachinelearning/webnn-native/pull/13/files#diff-8c735e422d57ddc0234da2ed876a76e5bf398a447c94c4e819b14f9d8fd7c812R889). @fujunwei

File: src/webnn_native/ops/ReduceMean.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        // the number of values in the sequence must be smaller than the rank of the input tensor
+        if (mOptions.axesCount > mRank) {
+            return DAWN_VALIDATION_ERROR("axes size is invalid.");
+        }
+
+        // the values in the sequence must be within the range from 0 to N-1
+        for (auto axis : mAxes) {
+            if (axis > int32_t(mRank - 1) || axis < -1) {
```
[the remark this answers is not in the document]
Comment (fujunwei) [reply src/webnn_native/ops/ReduceMean.cpp:57]: [static_cast](https://en.cppreference.com/w/cpp/language/static_cast)

**Comment by mingmingtasd** [comment]
Updated codes, PTAL, thanks! @fujunwei

**Review by fujunwei — APPROVED** [review]
LGTM.

File: src/webnn_native/openvino/ienn/src/ie_model.cpp [hunk 5]
```diff
@@ -438,6 +438,21 @@ ie_operand_t* Model::AddRelu(ie_operand_t* input) {
...
 }
 
+ie_operand_t* Model::AddReduceMean(ie_operand_t* input,
+                                   ie_reduce_mean_options_t* options) {
+  auto input_node = name_node_map_[input->name];
+  SizeVector axes = ToVector(options->axes, options->axesCount);
+  auto axes_node = op::Constant::create(element::i64, Shape{axes.size()}, axes);
+  // Issue: When input rank == 3 and keeDimensions == false, reduceMean produces
+  // a result with wrong shape. This will be fixed by
+  // https://github.com/openvinotoolkit/openvino/pull/5986 in OpenVINO 2021/4.
```
Comment (huningxin) [inline src/webnn_native/openvino/ienn/src/ie_model.cpp:448]: @mingmingtasd , could you please file an issue about this? So later we can upgrade to 2021/4 to fix this issue. Thanks!

**Review by huningxin — APPROVED** [review]
LGTM with a following-up. Please file a tracking issue, then we can merge. Thanks.

File: src/webnn_native/openvino/ienn/src/ie_model.cpp (hunk 5 above)
Comment (mingmingtasd) [reply src/webnn_native/openvino/ienn/src/ie_model.cpp:448]: Done, thanks! @huningxin

Comment (huningxin) [reply src/webnn_native/openvino/ienn/src/ie_model.cpp:448]: Could you please share the link? I mean https://github.com/webmachinelearning/webnn-native/issues

Comment (mingmingtasd) [reply src/webnn_native/openvino/ienn/src/ie_model.cpp:448]: Yes, I have raised two issues in this repo, #18 for IE and #19 for DML. Thanks! @huningxin

Comment (huningxin) [reply src/webnn_native/openvino/ienn/src/ie_model.cpp:448]: cool

**Comment by huningxin** [comment]
@mingmingtasd , please help resolve the conflicts, then we can merge. Later when #16 merged, we can avoid checking binary files.

**Comment by mingmingtasd** [comment]
> @mingmingtasd , please help resolve the conflicts, then we can merge. Later when #16 merged, we can avoid checking binary files.

Done, @fujunwei will rebase to remove binary files in #16 , PTAL and help merge, thanks! @huningxin

**Comment by huningxin** [comment]
Thanks!

**Pull request merged**